### ISSUE TITLE
support a fallback of using the DeltaChat folder similar to the past

### ISSUE THREAD

**Issue by r10s** [deltachat/deltachat-core#422]
the single-folder approach may have the potential to annoy exiting users that are used to the `DeltaChat` folder.

we should add a way to the core library that support the workflow of moving messages from the INBOX to another folder, so that in a shared-usage, the INBOX stays empty.

we could do this by using a 3rd thread that watches the INBOX (in addition to the [configured single-folder](https://github.com/deltachat/deltachat-core/pull/407)) and moves all messages in the INBOX with the `Chat-Version:` header to the folder configured as the single-folder.
i would not do much more as this would make things much more complicated (eg. folder creation (folders may be hidden, not hidden, subfolders of the INBOX and so so , this would require more maintainance) automatic upgrade and so on)

@angelo-fuchs on the android-ui i would suggest a checkmark in the advanced account settings, together with an option to type in the folder.

cc @hpk42

**Comment by testbird** [comment]
Thank you for re-considering to do some basic message sorting in the -core. Good to see you found the time for some real reasoning about this. As deltachat knows most about the message status, I think it is a very well suited place to do the sorting intelligently, providing a good overall user exprience.

I think it was @t-oster who first found out and formulated that the INBOX would always be the main folder to watch (https://github.com/deltachat/deltachat-core/issues/117#issuecomment-395687590). It allows to detect incoming (short chat) email messages, and replies from classic MUAs, and to move messages into a separate folder after chat messages have been displayed or email messages were answered in a chat (also allows to (intermittently) continue a chat from a classic email client, even with concurrently running multiple chat and email clients). While the "AltMarkMove" scheme (https://support.delta.chat/t/wiki-imap-strategy-to-live-within-the-email-ecosystem/123) also reduces the distubancees of email client listings (all chats immediately flaged "seen") and only requires a *single* folder watch.

If we leave out all the originally SIEVE ideas in the wiki, and the detection of "email-approved" (known contacts), the basic things do stay plain simple and manageable. But as the long list of isolated points may not have made it obvious how the concurrent syncing could work, and unfortunately nobody asked, I added the following explanation to the "basic approaches" section:

> By default clients follow all (INBOX) message updates seen on the server (i.e. additions and deletes) locally, and “upload” all local changes (i.e. deletes) to the server. To allow syncing multiple clients, there are some additional conditions. (1) To allow syncing multiple clients if only one (the INBOX) folder can be watched, before really locally deleting a message that was deleted from INBOX, wait a second and check if the message has not only been moved to the email chat subfolder by another client (means chat message has been displayed or email message has gotten a chat answer). (2) To still allow immediate propagation of a deletion in the email chat folder to all other clients, move deleted messages (IDs) back to the INBOX before deleting the message. And (3) have an option for clients to mark messages as “archived message” kept locally (i.e. do not delete locally if deleted on server, and ask about also deleting on server if the “archived” message is still present on the server).

**Comment by r10s** [comment]
@testbird the problems are complex and also include user expectations, which mails pop up and so on.

each approach has its advantages and disadvantates, many things are also a matter of defaults.

however, as you also have noticed, things are not carved in stone :)

**Comment by testbird** [comment]
Exactly, what you say was the starting point for all the considerations that let to the current state of the imap and the use-case wiki pages. Not all things are that obvious, but you'll probably come across more of them sooner or later anyway. If you would spell out any potential problem you may see I might be able to tell you if and how it is already covered in the wiki.

I think the current sorting in deltachat-core is only missing a couple of smaller things, feature wise, but it may possibly help you a lot to re-factor the code into separate the watching, status changing, and moving parts, just as you wrote in this issue above.

I hope you don't loose too much work spurting into non-viable short-cuts like requesting and supporting users and their contacts to use separate addresses. And that my additions above about the syncing with a single INBOX folder-watch are written in an understandable way.

**Comment by testbird** [comment]
IIUC the breakthrough in reliable idle message reception was also the result of a fruitful refactoring session. :-)

**Comment by r10s** [comment]
removed this from the single-folder milestone for now; we will see how things work out before starting optimizations on this.

it will probably take some weeks until single-folder is shipped, so, for now this is not the most urgent thing to do.

**Comment by testbird** [comment]
Sure.

> the potential to annoy exiting users

...is probably low. :)

If I exit, many things won't annoy me anymore, but I'll remain somehow sad until there are email compatible chat clients.

**Comment by angelo-fuchs** [comment]
> @angelo-fuchs on the android-ui i would suggest a checkmark in the advanced account settings, together with an option to type in the folder.

Yes, there should be an option to manually control that. I think the default behavior should look like this:

User sets up Deltachat on a device and connects to his mail. Then

```
1) IF -> There is a Deltachat Folder with deltachat messages present:
THEN -> Setup Twofolder (Inbox + Deltachat) mode
2) ELSE IF -> There is no Deltachat Folder AND Inbox is relatively empty (say less then 10 messages)
THEN -> Setup Onefolder mode
3) ELSE IF -> The Inbox has many messages
THEN -> Setup Twofolder (Inbox + Deltachat) mode
4) ELSE -> Setup Onefolder mode
```

Reasoning:
1) User already had Deltachat running "the old way", no need to diverge from Design of Least Surprise unless explicitly requested
2) User has probably setup a dedicated account, use it the most efficient way possible
3) User has probably setup Deltachat with a mailbox he already uses for other stuff as well, no need to clutter his INBOX
4) We don't know what kind of setup is present, so we fallback to the most efficient use. User has to manually switch to Twofolder

Things that I would pay no attention to in Autosetup:
- Potentially existing folders (other then "Deltachat") that could be used by some prior installation for Deltachat: If you are that advanced, you can probably just fix it yourself
- The existence of Deltachat messages in INBOX: We have no idea why they might be present and they would be moved to Deltachat folder with the Twofolder path

**Comment by angelo-fuchs** [comment]
My expectations on what the additional thread for Twofolder does:

1. Check in INBOX for messages with ``Chat-Version`` Header
2. Move messages to Deltachat Folder.
-> Error: Message not found: Ignore (some other instance probably moved it already)
-> Error: Folder Deltachat not found: Count errors, at count of 10 Minutes Autoswitch to Onefolder mode

**Comment by hpk42** [comment]
Hi Mor, 

my main concerns regarding two-folder complexity are:

- multi-device handling see eg
  https://github.com/deltachat/deltachat-core/issues/409
  which relates to the problem that IDLE (or "imap-push") 
  only operates on a single folder.  Maybe this is less
  of a problem with the new BCC-self logic as all sent
  messages will arrive to INBOX and thus be noted by Desktop
  and mobile if sent from the other. However, if we have two folders
  we will need to have two connections and two threads 
  to perform IDLE with each, as @r10s originally noted. 

- user-expectation mismatch problems whether a message 
  should be moved from INBOX to DeltaChat ...
  some expressed they only want a particular thread to
  move to DeltaChat, some are fine with moving all messages
  from an accepted contact to DeltaChat. Also in the old
  version there were bugs which we could not resolve as
  some messages remained in INBOX but should have gone to DeltaChat.

Maybe these are solvable and we can retain the "automatically move
chat messages to special folder" behaviour (for old users). 
But let's discuss further after desktop integrates single-folder 
and we gather some real-life experience. 

holger

**Comment by testbird** [comment]
> My expectations on what the additional thread for Twofolder does:
>
>   1. Check in INBOX for messages with Chat-Version Header
>   2. Move messages to Deltachat Folder.
>    -> Error: Message not found: Ignore (some other instance probably moved it already)
>    -> Error: Folder Deltachat not found: Count errors, at count of 10 Minutes Autoswitch to Onefolder mode

Note that the message moving "thread" should work better if it is done like postprocessing or garbage collection (moving messages and locally adjusting db location info, even if moved by another instance).

This allows to maintain flexibility in message moving (e.g. leave short classic MUA emails in INBOX until replied in a chat (implying they belong to the chat), or specific user preferences).

Another advantage is that the imap-idle thread could then always work in the same way, only watching the the INBOX and detecting the chat messages with most adequate fitering (`Chat-Version`, replies to chat messages, very short new emails, ...).

Actually, as all messages go through the INBOX, even for syncing multiple messages, there is no need for a second continuous imap-idle thread watching the deltachat subfolder. (See my quote from https://support.delta.chat/t/wiki-imap-strategy-to-chat-within-the-email-ecosystem/123 -> [above](https://github.com/deltachat/deltachat-core/issues/422#issuecomment-436632631).)

**Comment by testbird** [comment]
All running instances pick up messages as they arrive in INBOX, newly started instances check the deltachat subfolder once to get to the up-to-date message state.

**Comment by testbird** [comment]
Another advantage is that it would be trivial configuration option to enable/disable the message moving feature.

**Comment by r10s** [comment]
> Note that the message moving "thread" should work better if it is done like postprocessing or garbage collection (moving messages and locally adjusting db location info, even if moved by another instance).

what do you mean with "garbage collection" here?

**Comment by testbird** [comment]
That moving is not done with the same permanently-running priority than the imap-folder watch, rather "running behind" after the fact that the deltachat idle-watch detects message status changes and notifies the user.

I understood the idea presented by angelo-fuchs, as if that would require a moving-thread to first move all chat messages before a watch-thread would be able to pick it up. But I think that this dependency would limiting flexibility and make the setup and configuring unnecessary complicated.

**Comment by r10s** [comment]
@testbird okay. was just a bit confused about the wording "garbage collection".

in general, i had similar ideas a @angelo-fuchs, however, maybe without auto-setup, by just _offer_ a way to use the app as before. i would probably also not push new user to use the DeltaChat folder way. i also think the distinction via the number of mails in the inbox does not work that easy, eg. i know of some users that keep their inbox empty by moving things around.

it seem to be mostly a problem of user expectations that differ totally.

so, i think, whatever for complicated things we invest, there is always a reasonable amount of users going a different way. so maybe the "easiest" way for us does not annoy more users as if we do things in a very complicated way ;)

finally, i agree with @hpk42 that it makes sense to collect some experiences first. this, however, required single-folder to be in use on at least some devices.

**Comment by testbird** [comment]
Differences in user expectations and preferences can be solved with designs that are modular enough to allow for these adaptions (configuration options).

What's better?
Invest in complicated things to support an inflexible mode and closing the trap with a list of additional artificial dependencies and restrictions.

Or invest in a design that may seem complicated at first, but is actually less complicated because modular implementation follows (is directly analog) to the users interests (detecting, db storing, moving, archiving, ...) and maintains the full flexibility.

Well, it's your choice. Thinking, thinking again, before talking, and before following $talk.

**Comment by testbird** [comment]
Some more thoughts for multi-client @angelo-fuchs ideas about a thread that moves messages.
Trying to describe the idea better than "doing it like garbage collection". ;-)

It could make sense to let all clients rely on the INBOX to pick up on incoming messages reliably and ASAP. The moving could then be done with some seconds delay. This might also allow for better fault tolerance in the multiple client coordination.

The idle watch should be able trigger all clients, before one client does the moving operation and the others can just soft-fail their moving, or not even start.

* The collision of moving jobs may be reduced by letting the clients wait a random time, say between 3000-6000ms.
* If during the wait, the idle INBOX watch is triggered by another client already doing the move, a client deletes the local moving job for the message.
* If the wait completes with the job remaining, execute the job to move the message into the chat folder, but fail silently if the message isn't in the INBOX anymore, another parallel client was probably slightly faster.

**Comment by angelo-fuchs** [comment]
@testbird Will keep that in mind, thank you.

**Comment by testbird** [comment]
Your're welcome.

**Comment by csb0730** [comment]
As I read all these long comments: @testbird - very good description. Fully support your thoughts (incl. PS). Something similar I raised in earlier issues.

Too, I think that defering to move messages to chat folder could solve some problems in multi client szenarios and testbirds thoughts are showing easy ways to solve them.

For me it's not important to move messages with high priority to the Deltachat folder (1s or even 10min doesn't matter!). It only needs to be done if necessary in a reasonable time :-).

For explanation of my use case: My deltachat is always online by imap push, my standard MUA not. And only a common mail account.

**Comment by testbird** [comment]
Yes, doing the moving separately should make the code more maintainable, while still allowing to support nice special effects. For example, leaving short MUA messages (qualifying and displayed as chat) in the INBOX longer, until they actually get **replied** in deltachat. (Still allows to answer them with MUAs.)

**Comment by testbird** [comment]
[When to move messags from INBOX to DeltaChat folder?](https://github.com/deltachat/deltachat-core/issues/475)

**Comment by r10s** [comment]
closed by https://github.com/deltachat/deltachat-core/pull/478
